Mathlib/Geometry/Manifold/Sheaf/Smooth.lean

```diff
@@ -0,0 +1,403 @@
+import Mathlib.CategoryTheory.Sites.Whiskering
+import Mathlib.Geometry.Manifold.Sheaf.Basic
+import Mathlib.Geometry.Manifold.Algebra.SmoothFunctions
+
+/-! # The sheaf of smooth functions on a manifold
+
+The sheaf of `𝕜`-smooth functions from a manifold `M` to a manifold `N` can be defined as a sheaf of
+types using the construction `StructureGroupoid.LocalInvariantProp.sheaf` from the file
+`Mathlib.Geometry.Manifold.Sheaf.Basic`.  In this file we write that down (a one-liner), then do the
+work of upgrading this to a sheaf of [groups]/[abelian groups]/[rings]/[commutative rings] when `N`
+carries more algebraic structure.  For example, if `N` is `𝕜` then the sheaf of smooth functions
+from `M` to `𝕜` is a sheaf of commutative rings, the *structure sheaf* of `M`.
+
+## Main definitions
+
+* `smoothSheaf`: The sheaf of smooth functions from `M` to `N`, as a sheaf of types
+* `smoothSheaf.eval`: Canonical map onto `N` from the stalk of `smoothSheaf IM I M N` at `x`,
+  given by evaluating sections at `x`
+* `smoothSheafGroup`, `smoothSheafCommGroup`, `smoothSheafRing`, `smoothSheafCommRing`: The
+  sheaf of smooth functions into a [Lie group]/[abelian Lie group]/[smooth ring]/[smooth commutative
+  ring], as a sheaf of [groups]/[abelian groups]/[rings]/[commutative rings]
+* `smoothSheafCommRing.forgetStalk`: Identify the stalk at a point of the sheaf-of-commutative-rings
+  of functions from `M` to `R` (for `R` a smooth ring) with the stalk at that point of the
+  corresponding sheaf of types.
+* `smoothSheafCommRing.eval`: upgrade `smoothSheaf.eval` to a ring homomorphism when considering the
+  sheaf of smooth functions into a smooth commutative ring
+* `smoothSheafCommGroup.compLeft`: For a manifold `M` and a smooth homomorphism `φ` between
+  abelian Lie groups `A`, `A'`, the 'postcomposition-by-`φ`' morphism of sheaves from
+  `smoothSheafCommGroup IM I M A` to `smoothSheafCommGroup IM I' M A'`
+
+# Main results
+
+* `smoothSheaf.eval_surjective`: `smoothSheaf.eval` is surjective.
+* `smoothSheafCommRing.eval_surjective`: `smoothSheafCommRing.eval` is surjective.
+
+## TODO
+
+There are variants of `smoothSheafCommGroup.compLeft` for `GroupCat`, `RingCat`, `CommRingCat`;
+this is just boilerplate and can be added as needed.
+
+Similarly, there are variants of `smoothSheafCommRing.forgetStalk` and `smoothSheafCommRing.eval`
+for `GroupCat`, `CommGroupCat` and `RingCat` which can be added as needed.
+
+Currently there is a universe restriction: one can consider the sheaf of smooth functions from `M`
+to `N` only if `M` and `N` are in the same universe.  For example, since `ℂ` is in `Type`, we can
+only consider the structure sheaf of complex manifolds in `Type`, which is unsatisfactory. The
+obstacle here is in the underlying category theory constructions, and there is WIP (as of June 2023)
+to fix this.  See
+https://github.com/leanprover-community/mathlib/pull/19153
+and cross-references there.
```

Comment:
It would be nice to give a bit more detail here.

I looked at https://github.com/leanprover-community/mathlib/pull/19153 and there are some useful references but I didn't really feel this enhanced my understanding beyond "some parts of the category theory library are probably not sufficiently universe polymorphic".

Reply:
a7d51ddc756fe498640cb0a10069e2c735b6d10a updates to what I think is the current status here. @semorrison Could you look at that diff to see whether it's an accurate description?

Reply:
LGTM.